go/apps/api/routes/v2_keys_verify_key/handler.go

```diff
@@ -0,0 +1,261 @@
+		opts = append(opts, keys.WithCredits(DefaultCost))
+	}
+
+	if req.Ratelimits != nil {
+		opts = append(opts, keys.WithRateLimits(*req.Ratelimits))
+	} else {
+		// check auto applied ratelimits
+		opts = append(opts, keys.WithRateLimits(nil))
+	}
+
+	if req.Permissions != nil {
+		query, queryErr := convertPermissionsToQuery(*req.Permissions)
+		if queryErr != nil {
+			return queryErr
+		}
+
+		opts = append(opts, keys.WithPermissions(query))
+	}
+
+	err = key.Verify(ctx, opts...)
+	if err != nil {
+		return err
+	}
+
+	res.Data = openapi.KeysVerifyKeyResponseData{
+		Code:        key.ToOpenAPIStatus(),
+		Valid:       key.Valid,
+		Enabled:     ptr.P(key.Key.Enabled),
+		Name:        ptr.P(key.Key.Name.String),
+		Permissions: ptr.P(key.Permissions),
+		Roles:       ptr.P(key.Roles),
+		KeyId:       ptr.P(key.Key.ID),
+
+		Credits:    nil,
+		Expires:    nil,
+		Identity:   nil,
+		Meta:       nil,
+		Ratelimits: nil,
+	}
+
+	remaining := key.Key.RemainingRequests
+	if remaining.Valid {
+		res.Data.Credits = ptr.P(remaining.Int32)
+	}
+
+	if key.Key.Expires.Valid {
+		res.Data.Expires = ptr.P(key.Key.Expires.Time.UnixMilli())
+	}
+
+	if key.Key.Meta.Valid {
+		err = json.Unmarshal([]byte(key.Key.Meta.String), &res.Data.Meta)
+		if err != nil {
+			return fault.Wrap(err, fault.Code(codes.App.Internal.UnexpectedError.URN()),
+				fault.Internal("unable to unmarshal key meta"),
+				fault.Public("We encountered an error while trying to unmarshal the key meta data."),
+			)
+		}
+	}
+
+	if key.Key.IdentityID.Valid {
+		res.Data.Identity = &openapi.Identity{
+			ExternalId: key.Key.ExternalID.String,
+			Id:         key.Key.IdentityID.String,
+			Ratelimits: nil,
+			Meta:       nil,
+		}
+
+		for _, ratelimit := range key.GetRatelimitConfigs() {
+			if ratelimit.IdentityID == "" {
+				continue
+			}
+
+			res.Data.Identity.Ratelimits = append(res.Data.Identity.Ratelimits, openapi.RatelimitResponse{
+				AutoApply: ratelimit.AutoApply == 1,
+				Duration:  int64(ratelimit.Duration),
+				Id:        ratelimit.ID,
+				Limit:     int64(ratelimit.Limit),
+				Name:      ratelimit.Name,
+			})
+		}
+
+		if len(key.Key.IdentityMeta) > 0 {
+			err = json.Unmarshal(key.Key.IdentityMeta, &res.Data.Identity.Meta)
+			if err != nil {
+				return fault.Wrap(err, fault.Code(codes.App.Internal.UnexpectedError.URN()),
+					fault.Internal("unable to unmarshal identity meta"),
+					fault.Public("We encountered an error while trying to unmarshal the identity meta data."),
+				)
+			}
+		}
+	}
+
+	if len(key.RatelimitResults) > 0 {
+		ratelimitResponse := make([]openapi.VerifyKeyRatelimitData, 0)
+		for _, result := range key.RatelimitResults {
+			if result.Response == nil {
+				continue
+			}
+
+			exceeded := !result.Response.Success
+			ratelimitResponse = append(ratelimitResponse, openapi.VerifyKeyRatelimitData{
+				AutoApply: result.AutoApply,
+				Duration:  result.Duration.Milliseconds(),
+				Exceeded:  &exceeded,
```

Comment:
Boolean as nullable, smells like red flag, why is it a pointer?

Reply:
yeah it shouldn't will fix